### ISSUE TITLE
[.NET 7] OptionsConfigurationServiceCollectionExtensions.Configure regression

### ISSUE THREAD

**Issue by ypoul** [dotnet/runtime#80062]
### Description

The setter of a property is not triggered in `.NET 7` but is triggered in `.NET 6` while configuring an option.

### Reproduction Steps

Considering the following example:

```cs
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;

var configuration = new ConfigurationBuilder()
    .AddJsonFile("appsettings.json")
    .Build();

var services = new ServiceCollection()
    .Configure<Configuration>(configuration.GetSection("Section1"));

var provider = services.BuildServiceProvider();
var conf = provider.GetService<IOptions<Configuration>>()!.Value;
var origins = conf.FullyQualifiedOrigins; // Empty in .NET 7, ok in .NET 6


public class Configuration
{
    private HashSet<string> _origins;
    private HashSet<string> _fullyQualifiedOrigins;

    [Obsolete("This property is obsolete. Use Origins instead.")]
    public string Origin { get; set; }

    public HashSet<string> Origins
    {
        get
        {
            if (_origins == null)
            {
                _origins = new HashSet<string>();

                // Since we're depricating Origin we ease the transition to move the value automatically, unless its null
                if (Origin != null)
                {
                    _origins.Add(Origin);
                }
            }

            return _origins;
        }
        set
        {
            _origins = value;
            _fullyQualifiedOrigins = new HashSet<string>(value.Select(o => o.ToFullyQualifiedOrigin()), StringComparer.OrdinalIgnoreCase);
        }
    }

    public HashSet<string> FullyQualifiedOrigins
    {
        get => _fullyQualifiedOrigins ?? new HashSet<string>
            {
                Origin?.ToFullyQualifiedOrigin()
            };
        private set => _fullyQualifiedOrigins = value;
    }
}

public static class StringExtensions
{
    public static string ToFullyQualifiedOrigin(this string origin)
    {
        var uri = new Uri(origin);

        if (UriHostNameType.Unknown != uri.HostNameType)
            return uri.IsDefaultPort ? $"{uri.Scheme}://{uri.Host}" : $"{uri.Scheme}://{uri.Host}:{uri.Port}";

        return origin;
    }
}

```

appsettings.json:
```
{
  "Section1": {
    "Origins": [
      "https://localhost:5001"
    ]
  }
}
```

The setter of the Origins property is not triggered .NET 7.

### Expected behavior

FullyQualifiedOrigins should be filled.

### Actual behavior

FullyQualifiedOrigins is empty.

### Regression?

Yes in `.NET 7`.

### Configuration

Windows 10 21H2
.NET SDKs installed:
- 6.0.401 [C:\Program Files\dotnet\sdk]
- 7.0.101 [C:\Program Files\dotnet\sdk]

### Other information

Tested also with `Microsoft.Extensions.Options.ConfigurationExtensions` 8.0.0-alpha.1.22628.3 and it's the same behavior as .NET 7

**Comment by manandre** [comment]
To fix your issue, you can replace the type for the `Origins` property and its `_origins` backing field from `HashSet<string>` to `ISet<string>`. 
About the regression when the property is a concrete type (like `HashSet` here), I do not know whether it was officially supported.

**Comment by tarekgh** [comment]
Looks this is a duplicate of https://github.com/dotnet/runtime/issues/80052

**Comment by ypoul** [comment]
@tarekgh looks like we can't access to #80052

**Comment by tarekgh** [comment]
@ypoul this is interesting. I was able to access this issue until yesterday. I cannot see it anymore. I have no idea what has happened to this issue. Also, the author of this issue @gregaa123 is no longer accessible either. 

Anyway, I'll activate this one to ensure at least we have a tracking for the reported issue.

**Comment by tarekgh** [comment]
I confirmed this is duplicate of https://github.com/dotnet/runtime/issues/79904. We are going to have a fix for it in the next servicing release `7.0.3`.
